cpp/src/plasma/io/client_connection.h

```diff
@@ -0,0 +1,155 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PLASMA_COMMON_CLIENT_CONNECTION_H
+#define PLASMA_COMMON_CLIENT_CONNECTION_H
+
+#include <deque>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "plasma/io/basic_connection.h"
+
+namespace plasma {
+
+/// A generic type representing a client connection to a server. This typename
+/// can be used to write messages synchronously to the server.
+template <typename T>
+class ServerConnection : public io::BasicConnection<T> {
+ public:
+  std::shared_ptr<ServerConnection<T>> shared_from_this();
+
+  /// Allocate a new server connection.
+  ///
+  /// @param stream A reference to the server stream.
+  /// @return std::shared_ptr<ServerConnection>.
+  static std::shared_ptr<ServerConnection<T>> Create(T&& stream);
+
+  /// Write a message to the client.
+  ///
+  /// @param type The message type (e.g., a flatbuffer enum).
+  /// @param length The size in bytes of the message.
+  /// @param message A pointer to the message buffer.
+  /// @return Status.
+  std::error_code WriteMessage(int64_t type, int64_t length, const uint8_t* message);
+
+  /// Read a message from the client.
+  ///
+  /// @param type The message type (e.g., a flatbuffer enum).
```

Comment:
Let's switch all of the `@param` and `@return` to `\param` and `\return`, which is what is used in the rest of Arrow I believe.